File: server/src/main/java/com/willkamp/vial/api/Request.java
package com.willkamp.vial.api;

import java.util.Map;
import java.util.Optional;

public interface Request {
  Iterable<Map.Entry<CharSequence, CharSequence>> headers();

  Optional<String> getBodyText();

  Optional<String> pathParam(String key);

  <T> Optional<T> bodyJson(Class<T> clazz);
}

File: server/src/main/java/com/willkamp/vial/api/Request.kt
package com.willkamp.vial.api

import java.util.*
import java.util.function.Consumer
import kotlin.collections.Map.Entry

interface Request {
    val bodyText: String?
    val bodyBin: ByteArray?
    fun headers(): Iterable<Entry<CharSequence, CharSequence>>
    fun pathParam(key: String): String?
    fun queryParams(key: String): List<String>
    fun queryKeys(): Set<String>
    fun queryParam(key: String): String?
    fun queryParamOption(key: String): Optional<String> {
        return Optional.ofNullable(queryParam(key))
    }
    fun pathParamOption(key: String): Optional<String> {
        return Optional.ofNullable(pathParam(key))
    }
    fun <T> bodyJson(clazz: Class<T>): T?
    fun respondWith(responseBuilder: Consumer<ResponseBuilder>)
}
